### PR TITLE
[ConstraintSystem] Replace `Fix` with richer `ConstraintFix` interface

### PR DESCRIPTION
**Pull request by xedin**
**New `ConstraintFix` interface to replace old `Fix`**
  
 - Holds all required information to fix the problem e.g. types, member names, conformances  etc.
 - Has attached locator;
 - Has attached diagnostic logic (which allows to simplify fix application logic);
 - Allows us to remove all unnecessary/scattered logic/storage from constraint system;
 - Uses less memory because all additional storage is deallocated at the end of solver scope,
    unlike today where `ConstraintSystem` holds it until its lifetime is over.

### CODE REVIEW

**Comment by gregomni** [comment]
This is great and definitely going in the right direction. Are `ConstraintFix` and `FailureDiagnostic` always going to be in a one-to-one correspondence as they are now, with `FailureDiagnostic` being short-lived? If so, maybe it would make sense to have just one set of these classes instead of two (with the API like `ConstraintFix` here, where `diagnose()` gets passed contextual args, rather than `FailureDiagnostic` getting them in the initializer).

**Comment by xedin** [comment]
To be honest I am not sure yet for tightly coupled fix and diagnostic are, need to try to disagnose more complicated failures like contextual mismatch to figure that out...

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Looks like everything is green, going to wait for a while if anybody got more feedback.

File: test/Constraints/diagnostics.swift [hunk 1]
```diff
@@ -93,11 +93,11 @@ func f7() -> (c: Int, v: A) {
   return f6(g) // expected-error {{cannot convert return expression of type '(c: Int, i: A)' to return type '(c: Int, v: A)'}}
 }
 
-func f8<T:P2>(_ n: T, _ f: @escaping (T) -> T) {} // expected-note {{where 'T' = '(_, _)'}}
+func f8<T:P2>(_ n: T, _ f: @escaping (T) -> T) {}
 f8(3, f4) // expected-error {{argument type 'Int' does not conform to expected type 'P2'}}
 typealias Tup = (Int, Double)
 func f9(_ x: Tup) -> Tup { return x }
-f8((1,2.0), f9) // expected-error {{global function 'f8' requires that '(_, _)' conform to 'P2'}}
+f8((1,2.0), f9) // expected-error {{argument type 'Tup' (aka '(Int, Double)') does not conform to expected type 'P2'}}
```
Comment (slavapestov) [inline test/Constraints/diagnostics.swift]: Nice improvement!

**Review by slavapestov — APPROVED** [review]

**Review by rudkx — APPROVED** [review]
LGTM!

I love getting the non-main-path code out of the way to make the remainder easier to follow.

**Comment by xedin** [comment]
I'm going to extract the commit which deals with operators into separate PR and re-run tests, so this is closer to NFC.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Pull request merged**